src/trainers/KNNTrainer.js

```diff
@@ -14,96 +14,98 @@ import { logFirehoseMetric } from "../helpers/metrics";
+    optimalValues.kValue = bestK;
+    return optimalValues;
+  }
+
+  possibleKValues(state) {
+    let datasetSize = state.data.length;
     const smallDatasetSize = 10;
     const mediumDatasetSize = 100;
+
+    let kValues = [];
     const minimalK = 1;
     const smallK = 5;
     const defaultRegressionK = datasetSize < mediumDatasetSize
       ? minimalK
       : smallK;
     const defaultClassificationK = Math.round(datasetSize / 3);
+    const defaultK = isRegression(state)
+      ? defaultRegressionK
+      : defaultClassificationK;
 
-    let bestModel = null;
-    let bestPredictedLabels = [];
-    let bestK = -1;
-    let bestAccuracy = -1;
     if (state.accuracyCheckExamples.length > 0) {
       if (datasetSize <= smallDatasetSize && !isRegression(state)) {
-        this.knn = new KNN(state.trainingExamples, state.trainingLabels, {
-          k: datasetSize
-        });
-        bestModel = this.knn;
-        bestK = datasetSize;
+        kValues.push(datasetSize)
       } else if (isRegression(state)) {
-        this.knn = new KNN(state.trainingExamples, state.trainingLabels, {
-          k: defaultRegressionK
-        });
-        bestModel = this.knn;
-        bestK = defaultRegressionK;
+        kValues.push(defaultRegressionK)
       } else {
-        const kValues = this.calculatePossibleKValues();
-        kValues
+        kValues = this.calculatePotentialKValues(state)
           .filter(kValue => kValue <= state.trainingExamples.length)
-          .forEach(kValue => {
-            this.knn = new KNN(state.trainingExamples, state.trainingLabels, {
-              k: kValue
-            });
-            var model = this.knn;
-            const predictedLabels = this.batchPredict(
-              state.accuracyCheckExamples
-            );
-            const accuracy = this.getAccuracyPercent();
-            if (accuracy > bestAccuracy) {
-              bestAccuracy = accuracy;
-              bestK = kValue;
-              bestModel = model;
-              bestPredictedLabels = predictedLabels;
-            }
-          });
-        }
-      } else {
-      const defaultK = isRegression(state)
-        ? defaultRegressionK
-        : defaultClassificationK;
-      this.knn = new KNN(state.trainingExamples, state.trainingLabels, {
-        k: defaultK
-      });
-      bestModel = this.knn;
-      bestK = defaultK;
+      }
+    } else {
+      kValues.push(defaultK)
     }
-    store.dispatch(setKValue(bestK));
-    store.dispatch(setAccuracyCheckPredictedLabels(bestPredictedLabels));
-    store.dispatch(setTrainedModel(bestModel));
-
-    const state2 = store.getState();
-
-    logFirehoseMetric("train-model", state2);
+    return kValues;
+  }
 
-    const accuracy = getPercentCorrect(state2);
-    store.dispatch(
-      setHistoricResult(state2.labelColumn, state2.selectedFeatures, accuracy)
-    );
+  calculatePotentialKValues(state) {
+    let datasetSize = state.data.length;
```

Comment:
Nit: `trainingExamplesSize` and `possibleKValues` can both be consts because they aren't reassigned